Failure executing sqlite3 after bedde6d (at and preopen semantics)

**Describe the bug**

After bedde6d I'm getting some rather inscrutable bugs while testing [`github.com/ncruces/go-sqlite3`](https://github.com/ncruces/go-sqlite3).

I'm not entirely sure what's going on.

**To Reproduce**

```sh
git clone --branch crash https://github.com/ncruces/go-sqlite3
cd go-sqlite3
./build_deps.sh
go run ./cmd
```

This panics inside my own code with the following error:
```
panic: module "sqlite3-1" closed with exit_code(71)

goroutine 1 [running]:
github.com/ncruces/go-sqlite3.(*Conn).free(0xc0000e26c0, 0x10cc20)
	/Users/cruces/Desktop/go-sqlite3/conn.go:166 +0xa5
github.com/ncruces/go-sqlite3.Open({0x11ab821, 0x7}, 0x0, {0x0, 0x0})
	/Users/cruces/Desktop/go-sqlite3/conn.go:92 +0xa29
main.main()
	/Users/cruces/Desktop/go-sqlite3/cmd/main.go:11 +0x2f
exit status 2
```

It's this section of code:
https://github.com/ncruces/go-sqlite3/blob/7c3b679b7dcc8fe36cf87a59447ab1048bf1b4d0/conn.go#L86-L94

After calling `c.api.open.Call` (which does a bunch of WASI FS calls) any function call seems to return this error (I'm panicking `free` errors). I'm guessing some `libc` internal state somehow gets corrupted?

**Expected behavior**

The above repro should print this:
```
2023/01/12 15:08:14 sqlite3: disk I/O error
exit status 1
````

Or, after #1026, it should succeed in creating a `test.db` with some tables in it.

Just so this is easier to repro, I uploaded my `sqlite3.wasm` (it's not that big) to the `crash` branch. Also did the context change, and updated `go.mod` to wazero `main` (with 59076bccf58a491ab6a5df2bfb834e37602bfc73). Force pushed. 

From a clean checkout:
```sh
git clone --branch crash https://github.com/ncruces/go-sqlite3
cd go-sqlite3
go run ./cmd
```

I get a somewhat different log:
```
--> ._start.command_export()
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
<--
--> .malloc.command_export(8)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
<-- 1100736
--> .malloc.command_export(4)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
<-- 1100832
--> .sqlite3_open_v2.command_export(1100736,1100832,6,0)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
	==> wasi_snapshot_preview1.fd_fdstat_get(fd=3)
	<== (stat={filetype=DIRECTORY,fdflags=,fs_rights_base=,fs_rights_inheriting=},errno=ESUCCESS)
	==> wasi_snapshot_preview1.path_open(fd=3,dirflags=SYMLINK_FOLLOW,path=test.db,oflags=CREAT,fs_rights_base=,fs_rights_inheriting=,fdflags=)
	<== (opened_fd=4,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_seek(fd=4,offset=0,whence=0,result.newoffset=1048360)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.fd_read(fd=4,iovs=1048360,iovs_len=1)
	<== (nread=0,errno=ESUCCESS)
<-- 0
--> .free.command_export(1100832)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EINVAL)
	==> wasi_snapshot_preview1.proc_exit(rval=71)
--> .sqlite3_close.command_export(1100904)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat=,errno=EBADF)
	==> wasi_snapshot_preview1.fd_close(fd=4)
	<== errno=EBADF
<-- 0
panic: module "sqlite3-1" closed with exit_code(71)

goroutine 1 [running]:
github.com/ncruces/go-sqlite3.(*Conn).free(0xc0001ca360, 0x10cc20)
	/Users/cruces/Desktop/go-sqlite3/conn.go:166 +0xa5
github.com/ncruces/go-sqlite3.Open({0x127f3db, 0x7}, 0x0, {0x0, 0x0})
	/Users/cruces/Desktop/go-sqlite3/conn.go:92 +0xa31
main.main()
	/Users/cruces/Desktop/go-sqlite3/cmd/main.go:11 +0x2f
exit status 2
```

I find it weird that every call does a new `fd_prestat_get` for `fd`s 3 and 4? But I admit I haven't really studied these syscalls…

Then after the `sqlite3_open_v2` on the call to `free`, `fd_prestat_get` for `fd` 4 becomes `EINVAL` instead of `EBADF`, and `proc_exit` happens with exit code 71. This in turn causes the panic in my own code (I `panic` on errors in `malloc`/`free`).

--- a/imports/wasi_snapshot_preview1/fs.go
+++ b/imports/wasi_snapshot_preview1/fs.go
@@ -1370,7 +1370,7 @@ func preopenPath(fsc *sys.FSContext, dirFD uint32) (string, Errno) {
 	if f, ok := fsc.LookupFile(dirFD); !ok {
 		return "", ErrnoBadf // closed
 	} else if !f.IsPreopen {
-		return "", ErrnoInval
+		return "", ErrnoBadf
 	} else {
 		// TODO: multiple pre-opens
 		return fsc.FS().GuestDir(), ErrnoSuccess

--- a/imports/wasi_snapshot_preview1/fs_test.go
+++ b/imports/wasi_snapshot_preview1/fs_test.go
@@ -766,10 +766,10 @@ func Test_fdPrestatGet_Errors(t *testing.T) {
 			name:          "not pre-opened FD",
 			fd:            dirFD,
 			resultPrestat: 0, // valid offset
-			expectedErrno: ErrnoInval,
+			expectedErrno: ErrnoBadf,
 			expectedLog: `
 ==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
-<== (prestat=,errno=EINVAL)
+<== (prestat=,errno=EBADF)
 `,
 		},
 		{
@@ -890,10 +890,10 @@ func Test_fdPrestatDirName_Errors(t *testing.T) {
 			fd:            dirFD,
 			path:          validAddress,
 			pathLen:       pathLen,
-			expectedErrno: ErrnoInval,
+			expectedErrno: ErrnoBadf,
 			expectedLog: `
 ==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=4)
-<== (path=,errno=EINVAL)
+<== (path=,errno=EBADF)
 `,
 		},
 	}

